charger/nrggen2_json.go

```diff
@@ -0,0 +1,314 @@
+	// TODO: i am not sure if we can set current_set with decimals over the query parameter, so just in case i limit it to the integer value
+	uriWithQueryParams := fmt.Sprintf(
+		"%s?current_set=%.0f&charge_pause=%d&energy_limit=%d",
+		nrg.apiURL(gen2.ControlPath),
+		control.CurrentSet,
+		control.ChargePause,
+		control.EnergyLimit)
+
+	if withPhaseSwitch {
+		uriWithQueryParams = fmt.Sprintf("%s&phase_count=%d", uriWithQueryParams, control.PhaseCount)
+	}
+
+	req, err := request.New(http.MethodGet, uriWithQueryParams, nil, request.AcceptJSON)
+	if err != nil {
+		return err
+	}
+
+	var res gen2.Control
+
+	if err := nrg.DoJSON(req, &res); err != nil {
+		switch {
+		case res.Response != "":
+			return errors.New(res.Response)
+		case err != io.EOF:
+			return err
+		}
+	}
+
+	nrg.controlG.Reset()
+	nrg.valuesG.Reset()
+
+	return nil
+}
+
+// Status implements the api.Charger interface
+func (nrg *NRGKickGen2) Status() (api.ChargeStatus, error) {
+	res, err := nrg.valuesG.Get()
+	if err != nil {
+		return api.StatusNone, err
+	}
+
+	switch res.General.Status {
+	case "UNKNOWN":
+		return api.StatusNone, nil
+	case "STANDBY":
+		return api.StatusA, nil
+	case "CONNECTED":
+		return api.StatusB, nil
+	case "CHARGING":
+		return api.StatusC, nil
+	case "ERROR":
+		switch res.General.ErrorCode {
+		case "CP_SIGNAL_VOLTAGE_ERROR":
+			fallthrough
+		case "CP_SIGNAL_IMPERMISSIBLE":
+			return api.StatusE, fmt.Errorf("%s", res.General.ErrorCode)
+		default:
+			return api.StatusF, fmt.Errorf("%s", res.General.ErrorCode)
+		}
+	case "WAKEUP":
+		return api.StatusNone, nil
+	default:
+		return api.StatusNone, fmt.Errorf("unhandled status type")
+	}
+}
+
+// Enabled implements the api.Charger interface
+func (nrg *NRGKickGen2) Enabled() (bool, error) {
+	return nrg.enabled, nil
+}
+
+// Enable implements the api.Charger interface
+func (nrg *NRGKickGen2) Enable(enable bool) error {
+	res, err := nrg.controlG.Get()
+
+	if err != nil {
+		return err
+	}
+
+	if enable {
+		res.ChargePause = 0
+	} else {
+		res.ChargePause = 1
+	}
+
+	err = nrg.updateControl(res, false)
+
+	if err == nil {
+		nrg.enabled = enable
+	}
+
+	return err
+}
+
+// MaxCurrent implements the api.Charger interface
+func (nrg *NRGKickGen2) MaxCurrent(current int64) error {
+	if current < 6 {
+		return fmt.Errorf("allowed range: 6.0 - rated_current (16.0A / 32.0A)")
+	}
+
+	res, err := nrg.controlG.Get()
+
+	if err != nil {
+		return err
+	}
+
+	res.CurrentSet = float64(current)
+
+	return nrg.updateControl(res, false)
+}
+
+func (nrg *NRGKickGen2) GetMaxCurrent() (float64, error) {
+	res, err := nrg.controlG.Get()
+
+	if err != nil {
+		return 0, err
+	}
+
+	return res.CurrentSet, nil
+}
+
+var _ api.Meter = (*NRGKickGen2)(nil)
+
+// CurrentPower implements the api.Meter interface
+func (nrg *NRGKickGen2) CurrentPower() (float64, error) {
+	res, err := nrg.valuesG.Get()
+
+	if err != nil {
+		return 0, err
+	}
+
+	return res.Powerflow.TotalActivePower, err
```

Comment:
This will panic if powerflow is nil. Either guard or change the struct definition.